# Exploring Ebay Car Sales Data

## 1. Introduction

In this project, we will handle a dataset comprised of information about used cars, announced on the classifieds section of the German eBay website, *eBay Kleinanzeigen*.
The original dataset, scraped and uploaded to Kaggle by user orgesleka, is no longer available, but you can now find it [here](https://data.world/data-society/used-cars-data).

The data dictionary for this dataset is the following:
- `dateCrawled`: When this ad was first crawled. All field-values are taken from this date.
- `name`: Name of the car.
- `seller`: Whether the seller is private or a dealer.
- `offerType`: The type of listing
- `price`: The price on the ad to sell the car.
- `abtest`: Whether the listing is included in an A/B test.
- `vehicleType`: The vehicle Type.
- `yearOfRegistration`: The year in which the car was first registered.
- `gearbox`: The transmission type.
- `powerPS`: The power of the car in PS.
- `model`: The car model name.
- `kilometer`: How many kilometers the car has driven.
- `monthOfRegistration`: The month in which the car was first registered.
- `fuelType`: What type of fuel the car uses.
- `brand`: The brand of the car.
- `notRepairedDamage`: If the car has a damage which is not yet repaired.
- `dateCreated`: The date on which the eBay listing was created.
- `nrOfPictures`: The number of pictures in the ad.
- `postalCode`: The postal code for the location of the vehicle.
- `lastSeenOnline`: When the crawler saw this ad last online.


## 2. Aim of the project
Clean the data and analyze the used car listing.

In [1]:
# import the pandas and NumPy libraries

import pandas as pd
import numpy as np

# read the autos.csv file into pandas, and assign it to the variable name autos
autos=pd.read_csv('autos.csv',encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Observations

Right away, we can see that the following columns have missing (null) values:
- `vehicleType`
- `gearbox`
- `model`
- `fuelType`
- `notRepairedDamage`

We can also observe the columns `price` and `odometer ` are objects, which we will probably have to modify into a float type.

We can see that the names of the columns have lower- and uppercase letters and no spaces between them. This is known as [camelcase](https://en.wikipedia.org/wiki/Camel_case#:~:text=Camel%20case%20(sometimes%20stylized%20as,word%20starting%20with%20either%20case.). It will be easier for us to handle column names in the future if we convert them into [snakecase](https://en.wikipedia.org/wiki/Snake_case). It will also be helpful to rename some columns with a more descriptive title of its content (e.g. convert `price` into `price_dollars`).

## 3. Column Renaming

In [4]:
# print an array of the existing column names

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
# rename columns into snakecase

autos.rename(columns={'dateCrawled':'date_crawled', 'offerType':'offer_type', 'price':'price_dollars', 'abtest':'ab_test',
       'vehicleType':'vehicle_type', 'yearOfRegistration':'registration_year','powerPS':'power_ps', 'odometer':'odometer_km', 'monthOfRegistration':'registration_month', 'fuelType':'fuel_type',
       'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created', 'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code',
       'lastSeen':'last_seen'}, inplace=True)

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type price_dollars  ab_test vehicle_type  registration_year  \
0  privat    Angebot        $5,000  control          bus               2004   
1  privat    Angebot        $8,500  control    limousine               1997   
2  privat    Angebot        $8,990     test    limousine               2009   
3  privat    Angebot        $4,350  control   kleinwagen               2007   
4  privat    Angebot        $1,350     test        kombi               2003   

     gearbox  power_ps   model odometer_km  registration_month fuel_type  \
0    manuell       158  andere   150,000km                   3       lpg   
1  automatik       286     7er   150,000km                   6    benzin   
2    manuell       102    golf    70,000km                   7    benzin   
3  automatik        71  fortwo    70,000km                   6    benzin   
4    manuell         0   focus   150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now that we have converted the column names into a more managable and informative form, we will do some basic data exploration to determine what other cleaning tasks need to be done. We will start by: 

- Dropping any text columns where all or almost all values are the same, as they don't have useful information for analysis.
- Finding examples of numeric data stored as text, which can be cleaned and converted. Our initial analysis has already detected `price_dollars` and `odometer_km` as such examples.

In [7]:
# view min/max/median/mean etc.

autos.describe(include='all')

date_crawled         name  seller offer_type price_dollars  \
count                 50000        50000   50000      50000         50000   
unique                48213        38754       2          2          2357   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot            $0   
freq                      3           78   49999      49999          1421   
mean                    NaN          NaN     NaN        NaN           NaN   
std                     NaN          NaN     NaN        NaN           NaN   
min                     NaN          NaN     NaN        NaN           NaN   
25%                     NaN          NaN     NaN        NaN           NaN   
50%                     NaN          NaN     NaN        NaN           NaN   
75%                     NaN          NaN     NaN        NaN           NaN   
max                     NaN          NaN     NaN        NaN           NaN   

       ab_test vehicle_type  registration_year  gearbox      power_ps  model  \
count    50000        44905       50000.000000    47320  50000.000000  47242   
unique       2            8                NaN        2           NaN    245   
top       test    limousine                NaN  manuell           NaN   golf   
freq     25756        12859                NaN    36993           NaN   4024   
mean       NaN          NaN        2005.073280      NaN    116.355920    NaN   
std        NaN          NaN         105.712813      NaN    209.216627    NaN   
min        NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%        NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%        NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%        NaN          NaN        2008.000000      NaN    150.000000    NaN   
max        NaN          NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

From the table above, we can conclude:
- The `seller` and `offer_type` columns have only 2 unique values. Furthermore, the top value has a frequency of 49999, meaning that all but one row present this value.  Therefore, those columns contain virtually no useful information for analysis.
- The columns `price_dollars` and `odometer_km` are numeric data stored as text and need to be converted.
- The `nr_of_pictures` column is irrelevant to us, as all rows present only the 0.0 value.



## 4. Deleting irrelevant columns:
### 4.1 `seller`, `offer_type` and `nr_of_pictures`

In [8]:
# Drop the columns

autos.drop(columns=['seller','offer_type','nr_of_pictures'], inplace=True)

In [9]:
autos.shape

(50000, 17)


## 5. Text to numberic dtype conversion:
### 5.1 `price_dollars`

In [10]:
# view min/max/median/mean etc.

autos['price_dollars'].describe()

count     50000
unique     2357
top          $0
freq       1421
Name: price_dollars, dtype: object

In [11]:
# Remove any non-numeric characters.

autos['price_dollars'] = autos['price_dollars'].str.replace(',','').str.replace('$','')                                                                                 

In [12]:
# Convert the column to a numeric dtype.

autos['price_dollars']=autos['price_dollars'].astype('int')

In [13]:
autos['price_dollars']

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price_dollars, Length: 50000, dtype: int64

### 5.2 `odometer_km`

In [14]:
# view min/max/median/mean etc.

autos['odometer_km'].describe()

count         50000
unique           13
top       150,000km
freq          32424
Name: odometer_km, dtype: object

In [15]:
# Remove any non-numeric characters.

autos['odometer_km'] = autos['odometer_km'].str.replace(',','').str.replace('km','')                                                                                 


In [16]:
# Convert the column to a numeric dtype.

autos['odometer_km'] = autos['odometer_km'].astype('int')

In [17]:
autos['odometer_km']

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Lengt

## 6. Data exploration:

### 6.1 `price_dollars`

In [18]:
# See how many unique values are in the column

autos['price_dollars'].unique().shape

(2357,)

In [19]:
# view min/max/median/mean etc.

autos['price_dollars'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

### Observations

From this, we can already tell that there are some problems with the values in `price_dollars`:
- The min value is 0 dollars, which is unusual.
- The max value is 1e+8 dollars, which is clearly incorrect.
- The mean is 9840.0 dollars standard deviation is 481104.4 dollars, which is likely caused by the outlier discussed above.

In [20]:
# calculate the counts for each value. 

autos['price_dollars'].value_counts()

0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price_dollars, Length: 2357, dtype: int64

We can see that there are 1421 entries with a price of 0 dollars.

In [21]:
# calculate the values with their respective counts in ascending order

autos['price_dollars'].value_counts().sort_index(ascending= True)

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

There are an additional 156 entries with a set price of 1 dollar. This could be because sellers do not wish to set a price upfront and prefer to negotiate the price of the car in private. We will remove these entries from our analysis. 

Other values that stand out are prices equal and superior to 999990 dollars. In total, there are 14 entries in this range.
Given that we are working with a dataset of used cars, these prices are unreasonably high and drive our mean and standard deviation up. Therefore, we will consider them outliers and remove them from our analysis.

In [22]:
# remove outliers

cleaned_prices_autos=autos[autos["price_dollars"].between(2,999989)]

In [23]:
cleaned_prices_autos.shape

(48409, 17)

In [24]:
# view min/max/median/mean etc.

cleaned_prices_autos['price_dollars'].describe()

count     48409.000000
mean       5907.909707
std        9068.263463
min           2.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dollars, dtype: float64

In [25]:
# calculte the counts for each value

cleaned_prices_autos['price_dollars'].value_counts()

500       781
1500      734
2500      643
1200      639
1000      639
600       531
3500      498
800       498
2000      460
999       434
750       433
900       420
650       419
850       410
700       395
4500      394
300       384
2200      382
950       379
1100      376
1300      371
3000      365
550       356
1800      355
5500      340
350       335
1250      335
1600      327
1999      322
400       321
         ... 
30933       1
4655        1
8777        1
10888       1
14525       1
68300       1
4280        1
175000      1
3020        1
11240       1
4440        1
269         1
190000      1
22790       1
12860       1
3725        1
50900       1
3949        1
5998        1
1996        1
4111        1
2459        1
12395       1
34940       1
4239        1
173         1
205         1
410         1
4335        1
17799       1
Name: price_dollars, Length: 2345, dtype: int64

After cleaning the `price_dollars` column, we are left with **48409 entries**. Our mean price is now **5907.9 dollars**, which is lower that before, as is the standard deviation. The most frequent price for an used car is **500 dollars**, which appears in **781 entries (1.6%)**. 


### 6.2 `odometer_km`

In [26]:
# See how many unique values are in the column

cleaned_prices_autos['odometer_km'].unique().shape

(13,)

In [27]:
# view min/max/median/mean etc.

cleaned_prices_autos['odometer_km'].describe()

count     48409.000000
mean     125788.902890
std       39737.761014
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [28]:
# calculte the counts for each value

cleaned_prices_autos['odometer_km'].value_counts()

150000    31307
125000     5046
100000     2108
90000      1733
80000      1414
70000      1215
60000      1154
50000      1011
40000       815
5000        815
30000       779
20000       762
10000       250
Name: odometer_km, dtype: int64

The `odometer_km` column seems to have no major issues:
- The min value is 5000 km, present in 815 entries.
- The max value is **150000 km**, which is also the most frequent value, occuring in **31307 entries (64.7%)**.
- The mean is **125788.9 km**, with a standard deviation of      39737.8 km. 


Most entries have a value superior to 50000 km. Since we are working with used cars, this seems reasonable. 
Therefore, we will not remove any entries based on unrealistically high or low `odometer_km` values.

In [29]:
# make a copy of the dataframe under a shorter name

clean_autos = cleaned_prices_autos.copy()

## 6.3 dates and times

There are several columns that represent date values:
- `date_crawled`: When the ad was first crawled. Added by the crawler.
- `last_seen`: When the crawler saw the ad last online. Added by the crawler
- `ad_created`: The date on which the eBay listing was created. Created by the website.
- `registration_month`: The month in which the car was first registered. Created by the website.
- `registration_year`: The year in which the car was first registered. Created by the website.


In [30]:
clean_autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48409 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          48409 non-null object
name                  48409 non-null object
price_dollars         48409 non-null int64
ab_test               48409 non-null object
vehicle_type          43891 non-null object
registration_year     48409 non-null int64
gearbox               46113 non-null object
power_ps              48409 non-null int64
model                 45978 non-null object
odometer_km           48409 non-null int64
registration_month    48409 non-null int64
fuel_type             44438 non-null object
brand                 48409 non-null object
unrepaired_damage     39405 non-null object
ad_created            48409 non-null object
postal_code           48409 non-null int64
last_seen             48409 non-null object
dtypes: int64(6), object(11)
memory usage: 6.6+ MB


### Observations

As of now, the `date_crawled`, `ad_created` and `last_seen` columns are recognized as string types (object). The `registration_year` and `registration_month` columns, on the other hand, are stored as numeric data. 

We will convert the data stored in `date_crawled`, `ad_created` and `last_seen` into numeric data, so we can process it quantitatively.

### 6.3.1  `date_crawled`

In [31]:
clean_autos['date_crawled'].describe()

count                   48409
unique                  46739
top       2016-03-09 11:54:38
freq                        3
Name: date_crawled, dtype: object

As we can see above, the first 10 characters correspond to the date (e.g. 2016-03-12). Therefore, we can extract the date values and then generate a distribution using the `Series.value_counts()` command. 

In [32]:
# extract the date from the string

clean_autos['date_crawled']=clean_autos['date_crawled'].str[:10]

In [33]:
clean_autos['date_crawled'].head()

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object

In [34]:
clean_autos['date_crawled'].value_counts()

2016-04-03    1868
2016-03-20    1830
2016-03-21    1806
2016-03-12    1789
2016-03-14    1773
2016-04-04    1766
2016-03-07    1745
2016-04-02    1718
2016-03-28    1687
2016-03-19    1682
2016-03-15    1659
2016-03-29    1652
2016-04-01    1633
2016-03-30    1633
2016-03-08    1611
2016-03-09    1600
2016-03-22    1594
2016-03-11    1578
2016-03-23    1562
2016-03-26    1561
2016-03-10    1559
2016-03-31    1540
2016-03-17    1531
2016-03-25    1528
2016-03-27    1507
2016-03-16    1429
2016-03-24    1423
2016-03-05    1228
2016-03-13     758
2016-03-06     681
2016-04-05     633
2016-03-18     625
2016-04-06     153
2016-04-07      67
Name: date_crawled, dtype: int64

In [35]:
# calculate the distribution of values in percentages instead of counts

clean_autos['date_crawled'].value_counts(normalize=True, dropna=False)


2016-04-03    0.038588
2016-03-20    0.037803
2016-03-21    0.037307
2016-03-12    0.036956
2016-03-14    0.036625
2016-04-04    0.036481
2016-03-07    0.036047
2016-04-02    0.035489
2016-03-28    0.034849
2016-03-19    0.034746
2016-03-15    0.034270
2016-03-29    0.034126
2016-04-01    0.033733
2016-03-30    0.033733
2016-03-08    0.033279
2016-03-09    0.033052
2016-03-22    0.032928
2016-03-11    0.032597
2016-03-23    0.032267
2016-03-26    0.032246
2016-03-10    0.032205
2016-03-31    0.031812
2016-03-17    0.031626
2016-03-25    0.031564
2016-03-27    0.031131
2016-03-16    0.029519
2016-03-24    0.029395
2016-03-05    0.025367
2016-03-13    0.015658
2016-03-06    0.014068
2016-04-05    0.013076
2016-03-18    0.012911
2016-04-06    0.003161
2016-04-07    0.001384
Name: date_crawled, dtype: float64

In [36]:
# calculate the distribution of values in percentages and in ascending order

clean_autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)


2016-03-05    0.025367
2016-03-06    0.014068
2016-03-07    0.036047
2016-03-08    0.033279
2016-03-09    0.033052
2016-03-10    0.032205
2016-03-11    0.032597
2016-03-12    0.036956
2016-03-13    0.015658
2016-03-14    0.036625
2016-03-15    0.034270
2016-03-16    0.029519
2016-03-17    0.031626
2016-03-18    0.012911
2016-03-19    0.034746
2016-03-20    0.037803
2016-03-21    0.037307
2016-03-22    0.032928
2016-03-23    0.032267
2016-03-24    0.029395
2016-03-25    0.031564
2016-03-26    0.032246
2016-03-27    0.031131
2016-03-28    0.034849
2016-03-29    0.034126
2016-03-30    0.033733
2016-03-31    0.031812
2016-04-01    0.033733
2016-04-02    0.035489
2016-04-03    0.038588
2016-04-04    0.036481
2016-04-05    0.013076
2016-04-06    0.003161
2016-04-07    0.001384
Name: date_crawled, dtype: float64

As we can observe above, there are 48409 entries in `date_crawled`.
The frequencies for each value are quite distributed between all the different values. The range of dates starts on **2016-03-05** and ends on **2016-04-07**.
According to this column, most ads were crawled in **2016-04-03 (3.86%)**.

### 6.3.2 `ad_created`

In [37]:
clean_autos['ad_created'].describe()

count                   48409
unique                     76
top       2016-04-03 00:00:00
freq                     1880
Name: ad_created, dtype: object

In [38]:
# extract the date from the string

clean_autos['ad_created']=clean_autos['ad_created'].str[:10]

In [39]:
# calculate the distribution of values in percentages 

clean_autos['ad_created'].value_counts(normalize=True, dropna=False)


2016-04-03    0.038836
2016-03-20    0.037865
2016-03-21    0.037534
2016-04-04    0.036853
2016-03-12    0.036770
2016-03-14    0.035262
2016-04-02    0.035179
2016-03-28    0.034973
2016-03-07    0.034787
2016-03-29    0.034064
2016-03-15    0.034002
2016-04-01    0.033713
2016-03-19    0.033630
2016-03-30    0.033547
2016-03-08    0.033300
2016-03-09    0.033114
2016-03-11    0.032928
2016-03-22    0.032742
2016-03-26    0.032308
2016-03-23    0.032101
2016-03-10    0.031916
2016-03-31    0.031854
2016-03-25    0.031688
2016-03-17    0.031275
2016-03-27    0.031048
2016-03-16    0.030036
2016-03-24    0.029333
2016-03-05    0.022930
2016-03-13    0.017022
2016-03-06    0.015328
                ...   
2016-02-20    0.000041
2016-02-18    0.000041
2016-02-26    0.000041
2016-02-14    0.000041
2016-01-10    0.000041
2016-02-05    0.000041
2016-02-24    0.000041
2016-02-02    0.000041
2016-02-12    0.000041
2015-09-09    0.000021
2016-02-08    0.000021
2016-01-22    0.000021
2015-12-05 

In [40]:
# calculate the distribution of values in percentages and in ascending order

clean_autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033114
2016-03-10    0.031916
2016-03-11    0.032928
2016-03-12    0.036770
2016-03-13    0.017022
2016-03-14    0.035262
2016-03-15    0.034002
2016-03-16    0.030036
2016-03-17    0.031275
2016-03-18    0.013593
2016-03-19    0.033630
2016-03-20    0.037865
2016-03-21 

In the `ad_created` column, there are also 48409 entries. The range of dates from the  columns starts on **2015-06-11** and ends on **2016-04-07**. According to this column, the most popular day for ad creation was **2016-04-03 (3.88%)**.


### 6.3.3 `last_seen`

In [41]:
clean_autos['last_seen'].describe()

count                   48409
unique                  38377
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object

In [42]:
# extract the date from the string

clean_autos['last_seen']=clean_autos['last_seen'].str[:10]

In [43]:
# calculate the distribution of values in percentages

clean_autos['last_seen'].value_counts(normalize=True, dropna=False)


2016-04-06    0.221591
2016-04-07    0.132021
2016-04-05    0.124935
2016-03-17    0.028073
2016-04-03    0.025202
2016-04-02    0.024851
2016-03-30    0.024747
2016-04-04    0.024500
2016-03-31    0.023839
2016-03-12    0.023797
2016-04-01    0.022868
2016-03-29    0.022331
2016-03-22    0.021380
2016-03-28    0.020885
2016-03-20    0.020637
2016-03-21    0.020616
2016-03-24    0.019748
2016-03-25    0.019191
2016-03-23    0.018592
2016-03-26    0.016815
2016-03-16    0.016443
2016-03-15    0.015865
2016-03-19    0.015824
2016-03-27    0.015617
2016-03-14    0.012622
2016-03-11    0.012374
2016-03-10    0.010618
2016-03-09    0.009626
2016-03-13    0.008862
2016-03-08    0.007375
2016-03-18    0.007333
2016-03-07    0.005412
2016-03-06    0.004338
2016-03-05    0.001074
Name: last_seen, dtype: float64

In [44]:
# calculate the distribution of values in percentages and in ascending order

clean_autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)


2016-03-05    0.001074
2016-03-06    0.004338
2016-03-07    0.005412
2016-03-08    0.007375
2016-03-09    0.009626
2016-03-10    0.010618
2016-03-11    0.012374
2016-03-12    0.023797
2016-03-13    0.008862
2016-03-14    0.012622
2016-03-15    0.015865
2016-03-16    0.016443
2016-03-17    0.028073
2016-03-18    0.007333
2016-03-19    0.015824
2016-03-20    0.020637
2016-03-21    0.020616
2016-03-22    0.021380
2016-03-23    0.018592
2016-03-24    0.019748
2016-03-25    0.019191
2016-03-26    0.016815
2016-03-27    0.015617
2016-03-28    0.020885
2016-03-29    0.022331
2016-03-30    0.024747
2016-03-31    0.023839
2016-04-01    0.022868
2016-04-02    0.024851
2016-04-03    0.025202
2016-04-04    0.024500
2016-04-05    0.124935
2016-04-06    0.221591
2016-04-07    0.132021
Name: last_seen, dtype: float64

In the `last_seen` column, there are again 48409 entries. The range of dates from the  columns starts on **2016-03-05** and ends on **2016-04-07**. According to this column, the day when most ads were last seen online is **2016-04-06 (2.21%)**.

### Date and time: summary
        

| Column     | Range of dates | Most frequent day
| ----------- | ----------- | ----------- |
| **`date_crawled`**   | 2016-03-05 to 2016-04-07   |2016-04-03 (3.86%)|
| **`ad_created`**   | 2015-06-11 to 2016-04-07     |2016-04-03 (3.88%)|
| **`last_seen`**  | 2016-03-05 to 2016-04-07       |2016-04-06 (2.21%)|

### Main conclusions:

- `dateCrawled` displays the date the ads were crawled. Ad crawling started in March of 2016 and ended in April of the same year.
- Ads created as early as June 2015 are included in this dataset, as we can see in the `ad_created` column.
- The `last_seen` column contains the date when the crawler last saw the ad online. The day when most ads were last seen online is 2016-04-06 (2.21%).


## 6.4 'registration_year'

In [45]:
clean_autos['registration_year'].describe()

count    48409.000000
mean      2004.774319
std         88.783278
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As we can see above, the 'registration_year' column contains some errors. The maximum value is 9999 and the minimum is 1000, both clearly incorrect. 

Given that the 'registration_year' corresponds to the year in which the car was first registered, we known that cannot be after 2015, the year the first ads were created.
As for the earliest acceptable year for car registration, we will accept any year after 1900.

In [46]:
# remove the rows where the 'registation_year' values are outside the range we have defined

clean_autos = clean_autos[clean_autos['registration_year'].between(1900,2015)]

In [47]:
clean_autos['registration_year'].describe()

count    45322.000000
mean      2002.577291
std          6.924648
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       2015.000000
Name: registration_year, dtype: float64

In [48]:
clean_autos['registration_year'].value_counts(normalize=True)

2000    0.069017
2005    0.064582
1999    0.063678
2004    0.059574
2003    0.059508
2006    0.058912
2001    0.058073
2002    0.054720
1998    0.051895
2007    0.050174
2008    0.048828
2009    0.045960
1997    0.042849
2011    0.035810
2010    0.035060
1996    0.030096
2012    0.028904
1995    0.026764
2013    0.017696
2014    0.014607
1994    0.013878
1993    0.009355
2015    0.008627
1992    0.008120
1990    0.007546
1991    0.007480
1989    0.003773
1988    0.002957
1985    0.002074
1980    0.001809
          ...   
1966    0.000485
1976    0.000463
1969    0.000419
1975    0.000397
1965    0.000375
1964    0.000265
1963    0.000177
1959    0.000132
1961    0.000132
1956    0.000088
1958    0.000088
1937    0.000088
1910    0.000088
1962    0.000088
1955    0.000044
1951    0.000044
1934    0.000044
1950    0.000044
1954    0.000044
1957    0.000044
1941    0.000044
1953    0.000022
1948    0.000022
1929    0.000022
1943    0.000022
1939    0.000022
1938    0.000022
1927    0.0000

As we can see, the minimum value for this column is 1910, so cars on sale were registered as early as that. 
We can also observe that the year **2000** is the year in which the **highest number of cars were registered (6.9%)**. 

In [49]:
# select the top 15 most frequent values 

clean_autos['registration_year'].value_counts(normalize=True, dropna=False)[0:15]

2000    0.069017
2005    0.064582
1999    0.063678
2004    0.059574
2003    0.059508
2006    0.058912
2001    0.058073
2002    0.054720
1998    0.051895
2007    0.050174
2008    0.048828
2009    0.045960
1997    0.042849
2011    0.035810
2010    0.035060
Name: registration_year, dtype: float64

In [50]:
# add the frequencies of the top 15 most frequent values (excluding the first)

sum(clean_autos['registration_year'].value_counts(normalize=True, dropna=False)[0:15])

0.7986408366797583

In addition, almost **80%** of all car registrations occured between **1997** and **2011**. We can conclude from this analysis that, at the time of crawling, the majority of cars on sale in this platform were at least more than 5 years old. 


## 7. Determining `time_online`

Using the `last_seen` and `ad_created` columns, we can calculate how long on average an ad remains online until the car is sold. We will name this new variable `time_online`.


In [51]:
# convert 'ad_created' column to datetime

clean_autos['ad_created']=pd.to_datetime(clean_autos['ad_created'])

In [52]:
# convert 'last_seen' column to datetime

clean_autos['last_seen']=pd.to_datetime(clean_autos['last_seen'])

In [53]:
clean_autos['ad_created'].describe()

count                   45322
unique                     74
top       2016-04-03 00:00:00
freq                     1767
first     2015-06-11 00:00:00
last      2016-04-07 00:00:00
Name: ad_created, dtype: object

In [54]:
clean_autos['last_seen'].describe()

count                   45322
unique                     34
top       2016-04-06 00:00:00
freq                    10184
first     2016-03-05 00:00:00
last      2016-04-07 00:00:00
Name: last_seen, dtype: object

In [55]:
# create a new column with the time past between `ad_created` and `last_seen``

clean_autos['time_online']=clean_autos['last_seen']-clean_autos['ad_created']

In [56]:
clean_autos['time_online'].shape

(45322,)

In [57]:
# calculate the distribution of values in percentages 

clean_autos['time_online'].value_counts(normalize=True, dropna=False)

0 days      0.137373
2 days      0.103327
4 days      0.074313
1 days      0.056772
6 days      0.055889
3 days      0.049755
8 days      0.048012
9 days      0.036936
7 days      0.034001
11 days     0.033714
5 days      0.033648
10 days     0.026499
13 days     0.025705
12 days     0.024403
14 days     0.023344
15 days     0.021667
18 days     0.021424
16 days     0.019968
17 days     0.018380
22 days     0.017056
21 days     0.015754
19 days     0.012577
24 days     0.011915
23 days     0.011893
27 days     0.011473
29 days     0.011341
25 days     0.010767
26 days     0.010481
20 days     0.010326
28 days     0.008870
              ...   
40 days     0.000088
49 days     0.000066
34 days     0.000066
59 days     0.000066
55 days     0.000044
36 days     0.000044
43 days     0.000044
47 days     0.000044
53 days     0.000022
209 days    0.000022
39 days     0.000022
52 days     0.000022
300 days    0.000022
98 days     0.000022
51 days     0.000022
68 days     0.000022
41 days     0

In [58]:
# calculate the distribution of values in percentages and in ascending order

clean_autos['time_online'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

0 days      0.137373
1 days      0.056772
2 days      0.103327
3 days      0.049755
4 days      0.074313
5 days      0.033648
6 days      0.055889
7 days      0.034001
8 days      0.048012
9 days      0.036936
10 days     0.026499
11 days     0.033714
12 days     0.024403
13 days     0.025705
14 days     0.023344
15 days     0.021667
16 days     0.019968
17 days     0.018380
18 days     0.021424
19 days     0.012577
20 days     0.010326
21 days     0.015754
22 days     0.017056
23 days     0.011893
24 days     0.011915
25 days     0.010767
26 days     0.010481
27 days     0.011473
28 days     0.008870
29 days     0.011341
              ...   
37 days     0.000088
38 days     0.000110
39 days     0.000022
40 days     0.000088
41 days     0.000022
42 days     0.000022
43 days     0.000044
44 days     0.000022
45 days     0.000022
47 days     0.000044
48 days     0.000022
49 days     0.000066
51 days     0.000022
52 days     0.000022
53 days     0.000022
55 days     0.000044
58 days     0

According to this result, **13.9%** of ads were online for **less than a day**. It is likely that these ads were removed by the poster not because the car was sold, but because of some error in the description or price or something of that nature. 

In [59]:
# select the top 15 most frequent values (excluding the first)

clean_autos['time_online'].value_counts(normalize=True, dropna=False)[1:15]

2 days     0.103327
4 days     0.074313
1 days     0.056772
6 days     0.055889
3 days     0.049755
8 days     0.048012
9 days     0.036936
7 days     0.034001
11 days    0.033714
5 days     0.033648
10 days    0.026499
13 days    0.025705
12 days    0.024403
14 days    0.023344
Name: time_online, dtype: float64

In [60]:
# add the frequencies of the top 15 most frequent values (excluding the first)

sum(clean_autos['time_online'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)[1:15])

0.6263183442919554

### Main conclusions:

If we exclude the '0 days' value from the top 15 most frequent results, we obtain that **62.8%** of the ads are removed up to **14 days** after their creation. We can infer from this that most of the used cars announced in the  *eBay Kleinanzeigen* platform are sold after less than 2 weeks after publishing. 

## 8. Exploring mean price by brand

In [61]:
clean_autos['brand'].describe()

count          45322
unique            40
top       volkswagen
freq            9548
Name: brand, dtype: object

In [62]:
clean_autos['brand'].value_counts()

volkswagen        9548
bmw               5034
opel              4803
mercedes_benz     4411
audi              3956
ford              3163
renault           2109
peugeot           1343
fiat              1151
seat               820
skoda              754
nissan             692
mazda              688
smart              643
citroen            636
toyota             580
hyundai            457
sonstige_autos     440
volvo              419
mini               399
mitsubishi         370
honda              355
kia                327
alfa_romeo         302
porsche            279
suzuki             268
chevrolet          261
chrysler           161
dacia              121
daihatsu           115
jeep               104
subaru              97
land_rover          96
saab                76
jaguar              73
daewoo              69
trabant             65
rover               60
lancia              50
lada                27
Name: brand, dtype: int64

In [63]:
clean_autos['brand'].value_counts(normalize=True)

volkswagen        0.210670
bmw               0.111072
opel              0.105975
mercedes_benz     0.097326
audi              0.087287
ford              0.069790
renault           0.046534
peugeot           0.029632
fiat              0.025396
seat              0.018093
skoda             0.016637
nissan            0.015269
mazda             0.015180
smart             0.014187
citroen           0.014033
toyota            0.012797
hyundai           0.010083
sonstige_autos    0.009708
volvo             0.009245
mini              0.008804
mitsubishi        0.008164
honda             0.007833
kia               0.007215
alfa_romeo        0.006663
porsche           0.006156
suzuki            0.005913
chevrolet         0.005759
chrysler          0.003552
dacia             0.002670
daihatsu          0.002537
jeep              0.002295
subaru            0.002140
land_rover        0.002118
saab              0.001677
jaguar            0.001611
daewoo            0.001522
trabant           0.001434
r

We will start by selecting the top 10 most common brands. 

In [64]:
# select the top 10 brands

top_ten_brands=clean_autos['brand'].value_counts().index[0:10]
print(top_ten_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [65]:
# create an empty dictionary to hold the mean price for each brand

mean_price_brand={}

# loop over each top 10 brand and retrieve the mean price

for b in top_ten_brands:
    brand_rows=clean_autos[clean_autos['brand']==b]
    mean_price=brand_rows['price_dollars'].mean()
    mean_price_brand[b]=mean_price

In [66]:
mean_price_brand

{'audi': 9373.358442871588,
 'bmw': 8431.807310290027,
 'fiat': 2866.012163336229,
 'ford': 3824.3386025924756,
 'mercedes_benz': 8739.202448424394,
 'opel': 3029.383302102852,
 'peugeot': 3137.9471332836933,
 'renault': 2500.814129919393,
 'seat': 4439.796341463415,
 'volkswagen': 5493.5633640553}

### Main conclusions:

Among the top 10 most common brands, there are 3 **luxury brands**: `audi`, `mercedes_benz` and `bmw`. Used cars on sale by these brands have higher prices, costing on average around **9000 dollars**. 

The remaining brands are more affordable, and their mean prices range between 2500 to 5500 dollars. 

`fiat` and `renault` are the **cheapest brands** on average, with prices below **3000 dollars.** `volkswagen` and `seat` are **middle-range brands**, with mean prices closer to **5000 dollars**. 

## 9. Exploring mean price and average mileage 

In [67]:
# create an empty dictionary to hold the mean mileage for each brand

mean_miles_brand={}

# loop over each top 10 brand and retrieve the mean price

for b in top_ten_brands:
    brand_rows=clean_autos[clean_autos['brand']==b]
    mean_miles=brand_rows['odometer_km'].mean()
    mean_miles_brand[b]=mean_miles

In [68]:
mean_miles_brand

{'audi': 129246.71385237614,
 'bmw': 132446.36471990464,
 'fiat': 116589.92180712424,
 'ford': 123951.94435662345,
 'mercedes_benz': 130671.04964860575,
 'opel': 129244.22236102435,
 'peugeot': 127081.16157855547,
 'renault': 128013.27643432906,
 'seat': 120865.85365853658,
 'volkswagen': 128441.55844155845}

In [89]:
# convert the mean prices dictionary to a series object, using the series constructor

price_series=pd.Series(data=mean_price_brand)

In [90]:
# convert the mean mileage dictionary to a series object, using the series constructor

miles_series=pd.Series(data=mean_miles_brand)

In [91]:
price_series

audi             9373.358443
bmw              8431.807310
fiat             2866.012163
ford             3824.338603
mercedes_benz    8739.202448
opel             3029.383302
peugeot          3137.947133
renault          2500.814130
seat             4439.796341
volkswagen       5493.563364
dtype: float64

In [92]:
# create a dataframe from the price series using the dataframe constructor

price_miles_dataframe=pd.DataFrame(price_series, columns=['mean_price'])

In [93]:
price_miles_dataframe

mean_price
audi           9373.358443
bmw            8431.807310
fiat           2866.012163
ford           3824.338603
mercedes_benz  8739.202448
opel           3029.383302
peugeot        3137.947133
renault        2500.814130
seat           4439.796341
volkswagen     5493.563364

In [96]:
# add the mileage series as a new column in the new dataframe

price_miles_dataframe['mean_mileage']=miles_series

In [97]:
price_miles_dataframe

mean_price   mean_mileage
audi           9373.358443  129246.713852
bmw            8431.807310  132446.364720
fiat           2866.012163  116589.921807
ford           3824.338603  123951.944357
mercedes_benz  8739.202448  130671.049649
opel           3029.383302  129244.222361
peugeot        3137.947133  127081.161579
renault        2500.814130  128013.276434
seat           4439.796341  120865.853659
volkswagen     5493.563364  128441.558442

In [99]:
price_miles_dataframe['mean_mileage'].describe()

count        10.000000
mean     126655.206686
std        4835.453060
min      116589.921807
25%      124734.248662
50%      128227.417438
75%      129246.090980
max      132446.364720
Name: mean_mileage, dtype: float64

### Main conclusions:

- There are no drastic differences in the average mileage between the top 10 brands. All brands display a **mean mileage over 110000 km**, and the standard deviation of the series is below 5000 km.
- The brand with the highest average mileage is **`mercedes_benz` (132446.4 km)** and the one with the lowest is **`fiat` (116589.9 km)**. The 15856.5 km difference does not seem to justify the difference in mean prices, considering that `mercedes_benz` cars are 204.9% more expensive on average than `fiat` cars.
- Similarly, we can see that `opel` cars have almost exactly the same mean mileage as `audi` cars, while their average price is 67.7% cheaper.
- Therefore, we can conclude that the **average mileage of cars produced by the top 10 brands is not likely to influence the cars' average price**.

## 10. Final summary

We performed several data cleaning tasks before analyzing the dataset, such as:
- removing 3 irrelevant columns (`seller`, `offer_type`, and `nr_of_pictures`) that added no valuable information;
- excluding unreasonable entries in the `odometer_km`, `price_dollars` and `registration_year` columns, which we categorized as outliers.
    

Regarding the analysis of the data, we have determined that:
- at the time of crawling, the majority of cars on sale in the *eBay Kleinanzeigen* platform were at least more than 5 years old;
- most of the used cars announced in the crawled ads were sold after less than 2 weeks after publishing;
- Among the top 10 most common brands, there were 3 luxury brands (`audi`, `mercedes_benz` and `bmw`, average price ~9000 dollars), middle-range brands (`volkswagen` and `seat`, average price ~5000 dollars) and cheaper brands (`fiat` and `renault`, average price ~3000 dollars);
- The mean mileage of cars produced by the top 10 brands does not strongly correlate with the average price and it is unlikely to influence the cars' price.
